sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,263 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Azure.Messaging.EventHubs.Core
+{
+    /// <summary>
+    ///   A pool of <see cref="TransportProducer" />.
+    ///   It periodically checks if the producers are active and if not it disposes them.
+    /// </summary>
+    ///
+    internal class TransportProducerPool : IAsyncDisposable
+    {
+        /// <summary>The period in minutes after which <see cref="PerformExpiration"/> is run</summary>
+        private const int PerformExpirationPeriodAsMinutes = 10;
+
+        /// <summary>
+        ///   The set of active Event Hub transport-specific producers specific to a given partition;
+        ///   intended to perform delegated operations.
+        /// </summary>
+        ///
+        private ConcurrentDictionary<string, PoolItem> Pool { get; set; } = new ConcurrentDictionary<string, PoolItem>();
+
+        /// <summary>
+        ///   A reference to a <see cref="Timer"/> periodically checking the <see cref="TransportProducer"/> that
+        ///   are in use and those that can be closed.
+        /// </summary>
+        ///
+        private Timer ExpirationTimer { get; }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="TransportProducerPool"/> class.
+        /// </summary>
+        ///
+        public TransportProducerPool()
+        {
+            ExpirationTimer = new Timer(PerformExpiration(),
+                                        null,
+                                        TimeSpan.FromMinutes(PerformExpirationPeriodAsMinutes),
+                                        TimeSpan.FromMinutes(PerformExpirationPeriodAsMinutes));
+        }
+
+        /// <summary>
+        ///   It returns a <see cref="PooledProducer"/> corresponding to a given partition
+        ///   and it extends its sliding expiration.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The zero-based unique identifier of a partition associated with the Event Hub.</param>
+        /// <param name="connection">The <see cref="EventHubConnection"/> to use as the basis for delegation of client-type operations.</param>
+        /// <param name="retryPolicy">The <see cref="EventHubsRetryPolicy"/> to use as the basis for retrieving the information.</param>
+        ///
+        /// <returns>A <see cref="PooledProducer"/> corresponding to the partitionId passed in as parameter.</returns>
+        ///
+        /// <remarks>
+        ///   There is a slight probability that the returned producer may be closed at any time
+        ///   after it is returned and the caller may want to handle that scenario.
+        /// </remarks>
+        ///
+        public PooledProducer GetPartitionProducer(string partitionId,
```

Comment:
Hey @kinelski:  Can you give us another pair of eyes here?   Alberto and I have been brainstorming this for a bit and I'd love a fresh perspective as a sanity check...